### ISSUE TITLE
[Performance]  C# Gpu memory allocation

### ISSUE THREAD

**Issue by grinay** [microsoft/onnxruntime#13504]
### Describe the issue

Hello, we are building custom OCR system. we take PaddleOCR models, convert them to onnx format.
Now we are facing out of memory issue on GPU. 
I unable to find information about how to control  memory allocation on GPU in C# implementation. 
We have 4 models, one is 3mb, 2 models in size of 9 mb, and one big model 103mb.
We are using singleton pattern for Inference session and do not dispose them after each inference, 
```        
var options = new SessionOptions()
        {
            EnableMemoryPattern = false,
            EnableCpuMemArena = false,
            GraphOptimizationLevel = GraphOptimizationLevel.ORT_ENABLE_ALL,
            OptimizedModelFilePath =
                Path.Combine(Path.GetTempPath(), ModelPath.Replace(".onnx", ".optimized.onnx")),
        };

        if (OcrSettings.GpuDeviceId.HasValue)
        {
            var providerOptionsDict = new Dictionary<string, string>
            {
                ["device_id"] = OcrSettings.GpuDeviceId.Value.ToString(),
                ["arena_extend_strategy"] = "kSameAsRequested",
            };

            var cudaProviderOptions = new OrtCUDAProviderOptions();
            cudaProviderOptions.UpdateOptions(providerOptionsDict);
            options.AppendExecutionProvider_CUDA(cudaProviderOptions);
        }
```

after call var result= InferenceSession.Run(inputs); 
result is processed and disposed result.Dispose();
However after few inferences , we have Out of memory issue on GPU, nvidia-smi also shows that whole memory consumed.
Please advice how to deal with it. 

### To reproduce

Create singleton inference session which is not disposable

### Platform

Linux

### OS Version

Ubuntu 18.04

### ONNX Runtime Installation

Released Package

### ONNX Runtime Version or Commit ID

1.13.1

### ONNX Runtime API

C#

### Architecture

X64

### Execution Provider

CUDA

### Is this a quantized model?

Yes

**Comment by grinay** [comment]
Yes, however this doesn't affect anything. The issue is , any call to inference session run, will allocate memory for tensor and hold it until inference session object destroyed. However recreating inference session for each inference is expensive operation. We able to isolate memory issue only with destroying inference session after each run.  All the options which provided for CUDA, we tried, non of them solve that.

**Comment by yuslepukhin** [comment]
Onnxruntime employs caching CUDA memory allocations because they are more expensive than host memory allocations. For that reason, we pre-calculate memory patterns and try to re-use tensor memory allocations from one run to another. We are not optimizing for re-creating sessions for every run, it is very expensive. When you have dynamic shapes, it becomes more difficult to establish memory allocation patterns and one ends up using more memory. This is the issue number one. To deal with it you might experiment with disabling memory arena (not seeing it is exposed for CUDA though) and/or memory patterns (using `EnableMemoryPattern` property in SessionOptions)
This is likely to bring your performance down but trying does not hurt.

2) Another potential issue has to do with CUDA caching convolution memory because it has to keep its stuff there for faster execution and update it, which contributes to locking at one of the places.

None of the above has anything to do with C# though. On the C# side you want to make sure that you deal with disposable objects properly. Nearly all of the C# classes created on top of the native structures are disposable. This includes `SessionOptions`, `InferenceSession` and in particular, all of the output tensors that you get from the `Run()` call, they hold on to the native memory until you dispose of them. That could be CUDA memory, or it could be CPU memory, because by default ORT copies outputs to CPU so the client code can read them.

**Comment by yuslepukhin** [comment]
There is also an option to configure shrinking of the CUDA memory arena after every run.  [here](https://nam06.safelinks.protection.outlook.com/?url=https%3A%2F%2Fonnxruntime.ai%2Fdocs%2Fget-started%2Fwith-c.html&data=05%7C01%7Cdmitrism%40microsoft.com%7Ce8f79c01103547920db908dabe8c5a74%7C72f988bf86f141af91ab2d7cd011db47%7C1%7C0%7C638031807266903799%7CUnknown%7CTWFpbGZsb3d8eyJWIjoiMC4wLjAwMDAiLCJQIjoiV2luMzIiLCJBTiI6Ik1haWwiLCJXVCI6Mn0%3D%7C3000%7C%7C%7C&sdata=0bnUUzV60Jp%2BIdWt9RPQd49wv62%2BLaQTk6bigXVBYr8%3D&reserved=0)

However, it acquires a lock at some point, and it may interfere with concurrent runs on the same session object.

**Comment by genolve** [comment]
Also experiencing this problem of CUDA memory filling up for repeated calls to transformer model in Java. Following the advice on using memory arena shrinkage:
```
run_option = new OrtSession.RunOptions();
run_option.AddConfigEntry("kOrtRunOptionsConfigEnableMemoryArenaShrinkage", "gpu:0");
```

Does not work. As of version 1.13.1, run options has no method AddConfigEntry. It this still in development?
